lib/internal/modules/esm/loader.js

```diff
@@ -96,6 +100,59 @@ const {
 
 let emittedSpecifierResolutionWarning = false;
 
+function nextHookFactory(chain, meta, validate) {
+  // First, prepare the current
+  const { hookName } = meta;
+  const {
+    fn: hook,
+    url: hookFilePath,
+  } = chain[meta.hookIndex];
+
+  const nextHookName = `next${
+    StringPrototypeToUpperCase(hookName[0]) +
+    StringPrototypeSlice(hookName, 1)
+  }`;
+
+  // When hookIndex is 0, it's reached the default, which does not call next()
+  // so feed it a noop that blows up if called, so the problem is obvious.
+  const generatedHookIndex = meta.hookIndex;
+  let nextNextHook;
+  if (meta.hookIndex > 0) {
+    // Now, prepare the next: decrement the pointer so the next call to the
+    // factory generates the next link in the chain.
+    meta.hookIndex--;
+
+    nextNextHook = nextHookFactory(chain, meta, validate);
+  } else {
+    // eslint-disable-next-line func-name-matching
+    nextNextHook = function chainAdvancedTooFar() {
+      throw new RangeError(
+        `ESM custom loader '${hookName}' advanced beyond the end of the chain; this is a bug in Node.js itself.`
+      );
+    };
+  }
+
+  return ObjectDefineProperty(
+    async (...args) => {
+      // Update only when hook is invoked to avoid fingering the wrong filePath
+      meta.hookErrIdentifier = `${hookFilePath} '${hookName}'`;
+
+      validate(meta.hookErrIdentifier, ...args);
+
+      // Set when next<HookName> is actually called, not just generated.
+      if (generatedHookIndex === 0) { meta.chainFinished = true; }
+
+      const output = await hook(...args, nextNextHook);
```

Comment:
What happens when `hook` is sync, like if we make `resolve` sync?

Reply:
I thought of that after 😞 It might need to get re-un-DRY'd, but I think it's worth keeping this in case in future we go back to both being sync.

I think tomorrow's problem though? (I wrote the original code, and this, and will write the sync resolve, so history and blame will stay kosher)

Reply:
Sure, that’s fine. I assume there must be some way to write a branch like “if (unknown function) is async … else …”. Like `if (hook.constructor.name === 'AsyncFunction')` or something.

Reply:
There is :)